### ISSUE TITLE
Shell execution in DangerDSL

### ISSUE THREAD

**Issue by freak4pc** [danger/swift#347]
Hey all :) 

### tl;dr

Do you see it as a feasible idea (something you'd accept as a PR) to expose Shell Execution as part of `DangerDSL`?

### Full story

I finally gave up and started migrating away from danger-ruby and danger-swiftlint-ruby, which are now deprecated.

There are specific things in Swift that aren't easily expressible (or not as performant) as a simple bash command. For example, the following code which is inspired by an example from the old danger.systems website: 

```sh
# Make sure Quick focused test don't accidentally get merged
# and prevents all of our tests from running :O
fail("fit left in tests") if `grep -rI "fit(" Tests/`.length > 1
fail("fdescribe left in tests") if `grep -rI "fdescribe(" Tests/`.length > 1
fail("fcontext left in tests") if `grep -rI "fcontext(" Tests/`.length > 1
```

Doing this with NSFileManager is super expensive and verbose. I ended up making my own `Shell` so I can reproduce the above code in a decent way in Swift:

```swift
/// Make sure we didn't forget any focused Quick tests on
private func verifyNoQuickFocusedTests() {
    func testFiles(containing searchTerm: String) -> [String] {
        let result = (try? Shell.exec("grep -rlI \"\(searchTerm)\" Tests/")) ?? ""
        
        return result
            .components(separatedBy: "\n")
            .compactMap { piece -> String? in
                guard let spec = piece.components(separatedBy: "/")
                                      .last?
                                      .trimmingCharacters(in: .whitespacesAndNewlines),
                      !spec.isEmpty else { return nil }
                
                return spec
            }
    }
   
    let fitFiles = testFiles(containing: "fit(")
    let fdescribeFiles = testFiles(containing: "fdescribe(")
    let fcontextFiles = testFiles(containing: "fcontext(")
    
    if !fitFiles.isEmpty {
        fail("The following specs have fit() focused tests: \(fitFiles.joined(separator: ", "))")
    }

    if !fdescribeFiles.isEmpty {
        fail("The following specs have fdescribe() focused tests: \(fdescribeFiles.joined(separator: ", "))")
    }

    if !fcontextFiles.isEmpty {
        fail("The following specs have fcontext() focused tests: \(fcontextFiles.joined(separator: ", "))")
    }
}
```

To do this I made my own "Shell" extension which I thought could be a plugin similar to the SwiftLint plugin. Only ater I wrote this code, I noticed there's an internal `ShellExecutor`, but it's not accessible inside DangetDSL. 

```swift
public enum Shell {
    @discardableResult
    static func exec(_ command: String,
                     shell shellPath: String = "/bin/bash") throws -> String {
        let task = Process()
        task.launchPath = shellPath
        task.arguments = ["-c", command]

        let stdPipe = Pipe()
        let errPipe = Pipe()

        task.standardOutput = stdPipe
        task.standardError = errPipe
        task.launch()
        task.waitUntilExit()

        if task.terminationStatus == 0 {
            return stdPipe.output ?? ""
        } else {
            throw Error(exitCode: task.terminationStatus,
                        localizedDescription: errPipe.output ?? stdPipe.output ?? "Unknown error, exited with \(task.terminationStatus)")
        }
    }

    struct Error: Swift.Error {
        let exitCode: Int32
        let localizedDescription: String
    }
}

private extension Pipe {
    var output: String? {
        let output = fileHandleForReading.readDataToEndOfFile()

        guard let string = String(data: output, encoding: .utf8)?
                            .trimmingCharacters(in: .whitespacesAndNewlines),
              !string.isEmpty else { return nil }

        return string
    }
}
```

Thanks!

**Comment by f-meloni** [comment]
Hey @freak4pc :)
We have those two functions on `danger.utils`, that should be doing what you are suggesting (unless I misunderstood your needs)

https://github.com/danger/swift/blob/master/Sources/Danger/DangerUtils.swift#L58-L83

**Comment by freak4pc** [comment]
Ah ha! That was very well hidden :) 
I looked throughout the GitHub repo for something like "shell". 

Thanks @f-meloni!
